### PR TITLE
[FLINK-35016] catalog changes for model resource

### PR DESCRIPTION
**Pull request by lihaosky**
## What is the purpose of the change

Catalog changes to support model resource

## Brief change log

* Add `CatalogModel` related resource
* Add model CRUD operations in `CatalogManager` and `Catalog`

## Verifying this change

Unit test added

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by lihaosky** [comment]
cc @twalthr

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ResolvedCatalogModel.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    ResolvedSchema getResolvedInputSchema();
+
+    /** Returns a fully resolved and validated {@link ResolvedSchema} outputSchema. */
+    ResolvedSchema getResolvedOutputSchema();
+
+    /**
+     * Serializes this instance into a map of string-based properties.
+     *
+     * <p>Compared to the pure table options in {@link #getOptions()}, the map includes input
+     * schema, output schema, kind, task, comment and options.
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ResolvedCatalogModel.java:46]:
```suggestion
     * schema, output schema, comment and options.
```
kind and task are not important for the catalog layer

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ResolvedCatalogModel.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    Map<String, String> toProperties();
+
+    /**
+     * Creates an instance of {@link CatalogModel} from a map of string properties that were
+     * previously created with {@link ResolvedCatalogModel#toProperties()}.
+     *
+     * @param properties serialized version of a {@link ResolvedCatalogModel} that includes input
+     *     schema, output schema, kind, task, comment and options.
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ResolvedCatalogModel.java:55]:
```suggestion
     *     schema, output schema, comment and options.
```

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/CatalogPropertiesUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                new ResolvedSchema(columns, Collections.emptyList(), primaryKey);
+
+        final ResolvedCatalogTable testTable = new ResolvedCatalogTable(catalogTable, schema);
+
+        final Map<String, String> serializedMap =
+                CatalogPropertiesUtil.serializeCatalogTable(testTable);
+        final CatalogTable deserializedTable =
+                CatalogPropertiesUtil.deserializeCatalogTable(serializedMap);
+
+        // We need to compare the fields one by one since equalsTo in UnresolvedDataType in
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/CatalogPropertiesUtilTest.java:138]: nit: you can also simplify the code by comparing `schema.toString`. The toString is stable and used at a couple of places already. This should make this method very short.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ModelChange.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import java.util.Objects;
+
+/** {@link ModelChange} represents the modification of the model. */
+@PublicEvolving
+public interface ModelChange {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ModelChange.java:27]: as mentioned above, let's skip the ModelChange in this PR. It was not mentioned in the FLIP.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/ModelException.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * General Exception for all errors during model handling.
+ *
+ * <p>This exception indicates that an internal error occurred or that a feature is not supported
+ * yet. Usually, this exception does not indicate a fault of the user.
+ */
+@PublicEvolving
+public class ModelException extends RuntimeException {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/ModelException.java:30]: We also don't introduce dedicated exceptions for other catalog objects. Until there is an urgent need. Let's not add this.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 6]
```diff
@@ -1341,6 +1347,193 @@ private void dropTableInternal(
...
+
+    /**
+     * Like {@link #getModel(ObjectIdentifier)}, but throws an error when the model is not available
+     * in any of the catalogs.
+     */
+    public ResolvedCatalogModel getModelOrError(ObjectIdentifier objectIdentifier) {
+        return getModel(objectIdentifier)
+                .orElseThrow(
+                        () ->
+                                new ModelException(
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:1385]: Use same exception for tables: `TableException`. "Table" in this case refers not to the catalog object but this is a general exception.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 7]
```diff
@@ -1341,6 +1347,193 @@ private void dropTableInternal(
...
     }
 
+    /**
+     * Retrieves a fully qualified model. If the path is not yet fully qualified use {@link
+     * #qualifyIdentifier(UnresolvedIdentifier)} first.
+     *
+     * @param objectIdentifier full path of the model to retrieve
+     * @return model that the path points to.
+     */
+    public Optional<ResolvedCatalogModel> getModel(ObjectIdentifier objectIdentifier) {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:1357]: All catalog objects should also exist in a temporary version.

**Review by twalthr — COMMENTED** [review]
Thanks for the PR @lihaosky. I added my first set of comments. But the PR is too large for fast review cycles. Before I continue with the review:
It makes sense to remove the `ModelChanges` that were not mentioned in the FLIP. For now, we should only support `alterModel` with the final updated model.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java [hunk 8]
```diff
@@ -790,4 +792,126 @@ void alterPartitionColumnStatistics(
...
+     * same kind. For example, this doesn't allow altering a remote model to import model or native
+     * model, and vice versa. Note that the newModel contains only changes. Current alter model
+     * syntax supports alter properties and rename. This call is for altering model properties.
+     * Therefore, the newModel's properties are only changed properties. It's up to catalog
+     * implementation to apply the changes. The reason for this behavior is that this doesn't
+     * dictate how catalog implementation should handle model alteration. For example, it can do
+     * Read-modify-write or merge in place etc.
+     *
+     * @param modelPath path of the model to be modified
+     * @param modelChange the CatalogModel containing only changes
```
[the remark this answers is not in the document]
Comment (lihaosky) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java:903]: The original thought is that I don't want to check model existence like table in https://github.com/apache/flink/blob/master/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlNodeToOperationConversion.java#L420-L423 and return `NopOperation` if model doesn't exist. I want to let catalog check it and apply the change. But I guess I can introduce `ModelChange` like `TableChange` in `SqlModelAlterOperation` in next PR and make the model there optional if doesn't exist.

**Comment by twalthr** [comment]
@lihaosky can we close this PR then? It is very confusing if comments are spread across PRs. And I basically have to re-review everything because my "marked as reviewed" is lost.

**Comment by lihaosky** [comment]
Close in favor of https://github.com/apache/flink/pull/25211

**Pull request closed without merging**